src/Roslyn.Diagnostics.Analyzers/Core/NumberCommentslRefactoring.cs

```diff
@@ -0,0 +1,233 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Composition;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+
+namespace Roslyn.Diagnostics.Analyzers
+{
+    /// <summary>
+    /// This refactoring looks for numbered comments `// N` or `// N, N+1, ...` on non-empty lines
+    /// and checks that the numbers are sequential. If they are not, the refactoring is offered.
+    ///
+    /// This pattern is commonly used by compiler tests.
+    /// Comments that don't look like numbered comments are left alone. For instance, any comment that contains alpha characters.
+    /// </summary>
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(NumberCommentslRefactoring)), Shared]
+    class NumberCommentslRefactoring : CodeRefactoringProvider
```

Comment:
📝 `CodeRefactoringProvider` does not work via NuGet packages. This refactoring will be ignored by Visual Studio.